src/main/java/maksym/fedorenko/bookstore/service/impl/ShoppingCartServiceImpl.java

```diff
@@ -29,63 +29,74 @@ public class ShoppingCartServiceImpl implements ShoppingCartService {
-                ));
+        ShoppingCart cart = getUserShoppingCart(authentication);
         return shoppingCartMapper.toDto(cart);
     }
 
     @Override
-    public CartItemDto addCartItem(
+    @Transactional
+    public CartDto addCartItem(
             Authentication authentication, CreateCartItemRequestDto requestDto) {
-        String email = authentication.getName();
-        ShoppingCart cart = shoppingCartRepository.findByUserEmail(email)
-                .orElseThrow(() -> new EntityNotFoundException(
-                        "Can't find cart for user with email: %s".formatted(email)
-                ));
-        CartItem cartItem = cartItemRepository
-                .findByShoppingCartAndBookId(cart, requestDto.bookId())
-                .orElseGet(() -> createCartItem(requestDto, cart));
-        if (cartItem.getId() != null) {
-            cartItem.setQuantity(cartItem.getQuantity() + requestDto.quantity());
-        }
-        return cartItemMapper.toDto(cartItemRepository.save(cartItem));
+        checkIfBookExists(requestDto.bookId());
+        ShoppingCart cart = shoppingCartRepository.findByUserEmail(authentication.getName())
+                .orElseGet(() -> {
+                    ShoppingCart shoppingCart = new ShoppingCart();
+                    shoppingCart.setUser((User) authentication.getPrincipal());
+                    return shoppingCartRepository.save(shoppingCart);
+                });
+        cartItemRepository.findByShoppingCartAndBookId(cart, requestDto.bookId())
+                .ifPresentOrElse(
+                        item -> item.setQuantity(item.getQuantity() + requestDto.quantity()),
+                        () -> createCartItem(requestDto, cart)
+                );
+        return shoppingCartMapper.toDto(cart);
     }
 
     @Override
-    public CartItemDto updateCartItem(Long id, UpdateCartItemRequestDto requestDto) {
-        CartItem cartItem = findCartItemById(id);
+    @Transactional
+    public CartDto updateCartItem(
+            Authentication authentication, Long id, UpdateCartItemRequestDto requestDto) {
+        ShoppingCart cart = getUserShoppingCart(authentication);
+        CartItem cartItem = findCartItemByIdAndUser(authentication.getName(), id);
         cartItem.setQuantity(requestDto.quantity());
-        return cartItemMapper.toDto(cartItemRepository.save(cartItem));
+        return shoppingCartMapper.toDto(cart);
     }
 
     @Override
-    public void deleteCartItem(Long id) {
-        cartItemRepository.delete(findCartItemById(id));
+    public CartDto deleteCartItem(Authentication authentication, Long id) {
+        ShoppingCart cart = getUserShoppingCart(authentication);
+        CartItem cartItem = findCartItemByIdAndUser(authentication.getName(), id);
+        cart.removeCartItem(cartItem);
+        cartItemRepository.delete(cartItem);
+        return shoppingCartMapper.toDto(cart);
+    }
+
+    private ShoppingCart getUserShoppingCart(Authentication authentication) {
```

Comment:
can this be simplified to receive email?
```suggestion
    private ShoppingCart getUserShoppingCart(String email) {
```